File: Core/Workspace/Typescript/libs/workspace/adapters/memory/src/lib/Database/Result.cs
// <copyright file="RemoteResult.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json.Api;

    public abstract class Result : IResult
    {
        private readonly ISession session;
        private readonly Response response;

        private IDerivationError[] derivationErrors;

        protected Result(ISession session, Response response)
        {
            this.session = session;
            this.response = response;
        }

        public bool HasErrors => this.response.HasErrors;

        public string ErrorMessage => this.response.ErrorMessage;

        public IEnumerable<IObject> VersionErrors => this.session.Get<IObject>(this.response.VersionErrors);

        public IEnumerable<IObject> AccessErrors => this.session.Get<IObject>(this.response.AccessErrors);

        public IEnumerable<IObject> MissingErrors => this.session.Get<IObject>(this.response.MissingErrors);

        public IEnumerable<IDerivationError> DerivationErrors
        {
            get
            {
                if (this.derivationErrors != null)
                {
                    return this.derivationErrors;
                }

                if (this.response.DerivationErrors?.Length > 0)
                {
                    return this.derivationErrors ??= this.response.DerivationErrors
                        .Select(v => (IDerivationError)new DerivationError(this.session, v)).ToArray();
                }

                return this.derivationErrors;
            }
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/runtime/Result.ts
import { Response } from '@allors/protocol/json/system';
import { IDerivationError, IObject, IResult, ISession } from '@allors/workspace/domain/system';
import { DerivationError } from './DerivationError';

export abstract class Result implements IResult {
  private _derivationErrors: IDerivationError[];

  constructor(public readonly session: ISession, public readonly response: Response) {}

  get hasErrors(): boolean {
    return this.response._v?.length > 0 || this.response._a?.length > 0 || this.response._m?.length > 0 || this.response._d?.length > 0 || !!this.response._e;
  }

  get errorMessage(): string {
    return this.response._e;
  }

  get versionErrors(): IObject[] {
    return this.session.getMany(this.response._v);
  }

  get accessErrors(): IObject[] {
    return this.session.getMany(this.response._a);
  }

  get missingErrors(): IObject[] {
    return this.session.getMany(this.response._m);
  }

  get derivationErrors(): IDerivationError[] {
    if (this._derivationErrors != null) {
      return this._derivationErrors;
    }

    this._derivationErrors = this.response._d?.map((v) => new DerivationError(this.session, v)) ?? [];

    return this._derivationErrors;
  }
}
